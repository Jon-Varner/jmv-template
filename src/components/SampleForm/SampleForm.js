import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';

import TextField from '../UI/forms/TextField';
import Select from '../UI/forms/Select';
import CheckboxGroup from '../UI/forms/CheckboxGroup';
import RadioButtonGroup from '../UI/forms/RadioButtonGroup';

const SampleSchema = Yup.object().shape({
  fullName: Yup.string().required('Please enter your name.'),
  email: Yup.string().email('Please enter a valid email address.'),
  password: Yup.string()
    .min(8, 'Password must be at least 8 characters.')
    .required('Please enter a password.'),
  age: Yup.number()
    .required('Please enter your age.')
    .positive('Please enter a valid age.')
    .integer('Please enter a valid age.')
    .min(12, 'Age must be at least 12.')
    .max(120, 'Age must be lower than 120.')
});

const colorOptions = ['Red', 'Green', 'Blue', 'Yellow'];
const pronounOptions = ['he/him', 'she/her', 'they/them'];
const optinOptions = [
  { id: 'optedIn', label: 'yes', value: 'true' },
  { id: 'optedOut', label: 'no', value: 'false' }
];

export const SampleForm = ({ user, submitForm }) => (
  <Fragment>
    <h2>Current User Info:</h2>
    <ul className="returned-sample-state">
      {Object.keys(user).map(key => {
        let val = '';
        if (Array.isArray(user[key])) {
          val = user[key].join(', ');
        } else {
          val = user[key].toString();
        }

        return (
          <li key={key}>
            {key}: {val}
          </li>
        );
      })}
    </ul>

    <Formik
      initialValues={{
        fullName: '',
        email: '',
        password: '',
        age: '',
        preference: '',
        pronouns: ['they/them'],
        optin: 'false'
      }}
      validationSchema={SampleSchema}
      onSubmit={values => {
        submitForm(values);
      }}
      render={({ setFieldValue, setFieldTouched, values, errors, touched }) => (
        <Form>
          <TextField
            type="text"
            id="fullName"
            name="fullName"
            label="Name:"
            placeholder="Enter your name"
            required={true}
            errors={errors}
            touched={touched}
          />
          <TextField
            type="text"
            id="email"
            name="email"
            label="Email:"
            placeholder="Enter your email address"
            required={false}
            errors={errors}
            touched={touched}
          />
          <TextField
            type="password"
            id="password"
            name="password"
            label="Password:"
            placeholder="Enter your password"
            required={true}
            errors={errors}
            touched={touched}
          />
          <TextField
            type="number"
            id="age"
            name="age"
            label="Age:"
            placeholder="Enter your age"
            required={true}
            errors={errors}
            touched={touched}
          />
          <Select
            id="preference"
            name="preference"
            label="Color Preference:"
            placeholder="Select your preferred color"
            required={false}
            options={colorOptions}
            errors={errors}
            touched={touched}
          />
          <CheckboxGroup
            name="pronouns"
            label="Pronouns:"
            required={false}
            options={pronounOptions}
            errors={errors}
            touched={touched}
          />
          <RadioButtonGroup
            name="optin"
            label="Opt-in:"
            required={false}
            options={optinOptions}
            defaultValue={values.optin}
            errors={errors}
            touched={touched}
          />
          <button type="submit">Submit</button>
        </Form>
      )}
    />
  </Fragment>
);

SampleForm.propTypes = {
  user: PropTypes.object,
  submitForm: PropTypes.func.isRequired
};
