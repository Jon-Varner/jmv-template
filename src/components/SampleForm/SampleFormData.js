import uuid from 'uuid';
import * as yup from 'yup';

/*
  This data is used to dynamically generate form fields of each type.
*/

export const fields = [
  {
    name: 'name',
    id: uuid.v4(),
    type: 'text',
    label: 'Name:',
    required: 'true',
    placeholder: 'enter your name'
  },
  {
    name: 'email',
    id: uuid.v4(),
    type: 'email',
    label: 'Email:',
    required: 'false',
    placeholder: 'enter your email address'
  },
  {
    name: 'password',
    id: uuid.v4(),
    type: 'password',
    label: 'Password:',
    required: 'true',
    placeholder: 'enter a password'
  },
  {
    name: 'age',
    id: uuid.v4(),
    type: 'number',
    label: 'Age:',
    required: 'true',
    options: { min: 12, max: 120 }
  },
  {
    name: 'prefs',
    id: uuid.v4(),
    type: 'dropdown',
    label: 'Preference:',
    required: 'false',
    options: [
      { value: 1, displayValue: 'First' },
      { value: 2, displayValue: 'Second' },
      { value: 3, displayValue: 'Third' },
      { value: 4, displayValue: 'Fourth' }
    ]
  },
  {
    id: uuid.v4(),
    type: 'checkbox',
    label: 'Pronouns:',
    required: 'false',
    options: [
      { id: uuid.v4(), name: 'he', displayValue: 'he/him', checked: false },
      { id: uuid.v4(), name: 'she', displayValue: 'she/her', checked: false },
      { id: uuid.v4(), name: 'they', displayValue: 'they/them', checked: false }
    ]
  },
  {
    name: 'optin',
    id: uuid.v4(),
    type: 'radio',
    label: 'Subscribe:',
    required: 'false',
    options: [
      { id: uuid.v4(), value: true, displayValue: 'yes', checked: false },
      { id: uuid.v4(), value: false, displayValue: 'no', checked: false }
    ]
  }
];

/*
  This data is used to validate the dynamically generated form.
*/

export const validationSchema = yup.object().shape({
  name: yup.string().required('Please enter your name.'),
  email: yup.string().email('Please enter a valid email address.'),
  password: yup
    .string()
    .min(8, 'Password must be at least 8 characters.')
    .required('Please enter a password.'),
  age: yup
    .number()
    .required('Please enter your age.')
    .positive('Please enter a valid age.')
    .integer('Please enter a valid age.')
    .min(12, 'Age must be at least 12.')
    .max(120, 'Age must be lower than 120.')
});
