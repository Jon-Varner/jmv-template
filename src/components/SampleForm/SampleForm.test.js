import React from 'react';
import { mount } from 'enzyme';

import SampleForm from './SampleForm';

describe('SampleForm renders props correctly', () => {
  let props;
  let mountedSampleForm;

  const sampleForm = () => {
    if (!mountedSampleForm) {
      mountedSampleForm = mount(<SampleForm {...props} />);
    }
    return mountedSampleForm;
  };

  beforeEach(() => {
    props = {
      sampleData: 'test data',
      dispatchedSample: jest.fn()
    };

    mountedSampleForm = undefined;
  });

  it('always renders a paragraph and input', () => {
    const para = sampleForm().find('p');
    const input = sampleForm().find('input');

    expect(para.length).toBe(1);
    expect(input.length).toBe(1);
  });

  it('renders sampleData from props', () => {
    const para = sampleForm()
      .find('p')
      .first();

    const input = sampleForm()
      .find('input')
      .first();

    expect(para.text()).toBe('test data');
    expect(input.props().value).toBe('test data');
  });

  it('calls sampleData dispatch function on input change', () => {
    sampleForm()
      .find('input')
      .first()
      .simulate('change', {
        target: {
          value: 'new data'
        }
      });

    expect(props.dispatchedSample).toBeCalledWith('new data');
  });
});
