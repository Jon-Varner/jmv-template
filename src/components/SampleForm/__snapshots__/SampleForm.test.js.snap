// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form tests renders 1`] = `
<DocumentFragment>
  <h2>
    Current User Info:
  </h2>
  <ul
    class="returned-sample-state"
  >
    <li>
      fullName: Initial User Name
    </li>
    <li>
      email: Initial Email
    </li>
    <li>
      password: Initial Password
    </li>
    <li>
      age: 30
    </li>
    <li>
      preference: None
    </li>
    <li>
      pronouns: 
    </li>
    <li>
      optin: false
    </li>
  </ul>
  <form>
    <div
      class="form-field"
    >
      <label
        for="fullName"
      >
        Name: 
        <span
          class="required"
        >
          *
        </span>
      </label>
      <input
        id="fullName"
        name="fullName"
        placeholder="Enter your full name"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-field"
    >
      <label
        for="email"
      >
        Email: 
      </label>
      <input
        id="email"
        name="email"
        placeholder="Enter your email address"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-field"
    >
      <label
        for="password"
      >
        Password: 
        <span
          class="required"
        >
          *
        </span>
      </label>
      <input
        id="password"
        name="password"
        placeholder="Enter your password"
        type="password"
        value=""
      />
    </div>
    <div
      class="form-field"
    >
      <label
        for="age"
      >
        Age: 
        <span
          class="required"
        >
          *
        </span>
      </label>
      <input
        id="age"
        name="age"
        placeholder="Enter your age"
        type="number"
        value=""
      />
    </div>
    <div
      class="form-field"
    >
      <label
        for="preference"
      >
        Color Preference: 
      </label>
      <select
        id="preference"
        name="preference"
      >
        <option
          value="0"
        >
          Select your preferred color
        </option>
        <option
          value="Red"
        >
          Red
        </option>
        <option
          value="Green"
        >
          Green
        </option>
        <option
          value="Blue"
        >
          Blue
        </option>
        <option
          value="Yellow"
        >
          Yellow
        </option>
      </select>
    </div>
    <div
      class="form-field"
    >
      <h2>
        Pronouns: 
      </h2>
      <ul>
        <li>
          <input
            id="he/him"
            type="checkbox"
          />
          <label>
            he/him
          </label>
        </li>
        <li>
          <input
            id="she/her"
            type="checkbox"
          />
          <label>
            she/her
          </label>
        </li>
        <li>
          <input
            checked=""
            id="they/them"
            type="checkbox"
          />
          <label>
            they/them
          </label>
        </li>
      </ul>
    </div>
    <div
      class="form-field"
    >
      <h2>
        Opt-in: 
      </h2>
      <ul
        name="optin"
        value="false"
      >
        <li>
          <input
            id="optedIn"
            name="optin"
            type="radio"
            value="true"
          />
          <label
            for="optedIn"
          >
            yes
          </label>
        </li>
        <li>
          <input
            checked=""
            id="optedOut"
            name="optin"
            type="radio"
            value="false"
          />
          <label
            for="optedOut"
          >
            no
          </label>
        </li>
      </ul>
    </div>
    <button
      class="formSubmit"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
